### PR TITLE
Add support for unit-based duration parsing to AmountFormats

### PR DESCRIPTION
**Pull request by TristonianJones**
Introduce unit-based duration support based on Golang's `time.ParseDuration` method. 

It seems that the abseil.io library only supports parsing of these unit-based duration strings, but not encoding. If unit-based encoding is desired in Java, then Golang's `time.Duration.String()` implementation should serve as the reference implementation.

See #180

### CODE REVIEW

**Comment by TristonianJones** [comment]
@jodastephen Any concerns with the proposed submission for #180?

File: src/main/java/org/threeten/extra/AmountFormats.java [hunk 1]
```diff
@@ -287,6 +307,186 @@ private static boolean oppositeSigns(int a, int b) {
         return a < 0 ? (b >= 0) : (b < 0);
     }
 
+    // -------------------------------------------------------------------------
+    /**
+     * Parses a unit-based duration string to a duration value.
+     * <p>
+     * This is an implementation of Golang's {@link https://golang.org/pkg/time/#ParseDuration}:
```
Comment (jodastephen) [inline src/main/java/org/threeten/extra/AmountFormats.java:314]: Maybe rephrase something more like "The parses formatted durations based on units. The behaviour matches that of the <a href="https://golang.org/pkg/time/#ParseDuration">Golang</a> duration parser."

File: src/main/java/org/threeten/extra/AmountFormats.java [hunk 2]
```diff
@@ -287,6 +307,186 @@ private static boolean oppositeSigns(int a, int b) {
...
+     * unit suffix, such as "300ms", "-1.5h" or "2h45m". Valid time units are
+     * "ns", "us" (or "µs"), "ms", "s", "m", "h".
+     * <p>
+     * Note, the value "0" is specially supported as {@code Duration.ZERO}, but
+     * unlike the Go implementation infinite durations are not supported.
+     *
+     * @param durationText the formatted unit-based duration string.
+     * @return the {@code Duration} value represented by the string, if possible.
+     */
+    public static Optional<Duration> parseUnitBasedDuration(CharSequence durationText) {
```
Comment (jodastephen) [inline src/main/java/org/threeten/extra/AmountFormats.java:326]: The problem with returning optional is that no feedback is provided as to why the input is invalid. What was your thinking here? java.time.* generally throws exceptions in parsing.

File: src/main/java/org/threeten/extra/AmountFormats.java (hunk 1 above)
Comment (TristonianJones) [reply src/main/java/org/threeten/extra/AmountFormats.java]: Done.

File: src/main/java/org/threeten/extra/AmountFormats.java (hunk 2 above)
Comment (TristonianJones) [reply src/main/java/org/threeten/extra/AmountFormats.java]: I was copying the `absl::optional` return behavior of the C++ implementation, but have switched to the more Java idiomatic exception behavior by throwing `java.time.DateTimeParseException` for invalid inputs.

**Comment by TristonianJones** [comment]
@jodastephen PTAL

**Review by jodastephen — APPROVED** [review]
